tests/net/mod.rs

```diff
@@ -0,0 +1,796 @@
+    #[inline]
+    pub fn outputs(&self) -> &[D::Output] {
+        self.outputs.as_slice()
+    }
+}
+
+/// A network message on the virtual network.
+// Note: We do not use `messaging::TargetedMessage` and `messaging::SourceMessage` here, the nesting
+//       is inconvenient and we do not want to support broadcasts at this level.
+#[derive(Clone, Debug)]
+pub struct NetworkMessage<M, N> {
+    /// Message sender.
+    from: N,
+    /// Destined receiver.
+    to: N,
+    /// The actual message contents.
+    payload: M,
+}
+
+impl<M, N> NetworkMessage<M, N> {
+    /// Create a new network message.
+    #[inline]
+    fn new(from: N, payload: M, to: N) -> NetworkMessage<M, N> {
+        NetworkMessage { from, to, payload }
+    }
+}
+
+/// Mapping from node IDs to actual node instances.
+pub type NodeMap<D> = collections::BTreeMap<<D as DistAlgorithm>::NodeUid, Node<D>>;
+
+/// A virtual network message tied to a distributed algorithm.
+pub type NetMessage<D> =
+    NetworkMessage<<D as DistAlgorithm>::Message, <D as DistAlgorithm>::NodeUid>;
+
+/// Process a step.
+///
+/// Expands every message in the step by turning all broadcast messages into peer-to-peer messages,
+/// and appends them to the network queue. Additionally, saves a copy of each output to the output
+/// buffer of the `sender` node.
+///
+/// At the end, the number of additional messages created by non-faulty nodes is returned.
+///
+/// # Panics
+///
+/// The function will panic if the `sender` ID is not a valid node ID in `nodes`.
+// This function is defined outside `VirtualNet` and takes arguments "piecewise" to work around
+// borrow-checker restrictions.
+#[cfg_attr(feature = "cargo-clippy", allow(needless_pass_by_value))]
+fn process_step<'a, D>(
+    nodes: &'a mut collections::BTreeMap<D::NodeUid, Node<D>>,
+    sender: D::NodeUid,
+    step: &Step<D>,
+    dest: &mut collections::VecDeque<NetMessage<D>>,
+) -> usize
+where
+    D: DistAlgorithm + 'a,
+    D::Message: Clone,
+    D::Output: Clone,
+{
+    // For non-faulty nodes, we count the number of messages.
+    let faulty = nodes
+        .get(&sender)
+        .expect("Trying to process a step with non-existing node ID")
+        .is_faulty();
+    let mut message_count: usize = 0;
+
+    // Queue all messages for processing.
+    for tmsg in &step.messages {
+        match &tmsg.target {
+            // Single target message.
+            messaging::Target::Node(to) => {
+                if !faulty {
+                    message_count = message_count.saturating_add(1);
+                }
+
+                dest.push_back(NetworkMessage::new(
+                    sender.clone(),
+                    tmsg.message.clone(),
+                    to.clone(),
+                ));
+            }
+            // Broadcast messages get expanded into multiple direct messages.
+            messaging::Target::All => for to in nodes.keys() {
+                if *to == sender {
```

Comment:
Maybe add `filter` instead to the loop iterator for readibility?

Reply:
Sounds like a great idea. I'll add it.